# Part 1. Equation of a Slime

How many late days are you using for this assignment? 0

In [73]:
# Imports section
from sklearn.model_selection import train_test_split, cross_val_score, KFold
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score, mean_squared_error
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import PolynomialFeatures, StandardScaler

import pandas as pd

## 1. Loading the dataset

In [74]:
# Using pandas load the dataset
df = pd.read_csv('science_data_large.csv')
# Output the first 15 rows of the data
print(df.head(15),'\n')
# Display a summary of the table information (number of datapoints, etc.)
print(df.info())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000

## 2. Splitting the dataset

In [75]:
# Take the pandas dataset and split it into our features (X) and label (y)
x = df[['Temperature °C', 'Mols KCL']]
y = df['Size nm^3']

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.1, random_state=42)

## 3. Perform a Linear Regression

In [76]:
# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(X_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
sample_data = pd.DataFrame({
    'Temperature °C': [600],
    'Mols KCL': [500]
})
predicted_size = model.predict(sample_data)
formatted_predicted_size = f"{predicted_size[0]:.5g}"

print("Predicted Size nm^3 for Temperature = 600 °C & Mols KCL = 500:", formatted_predicted_size)

# Report on the score for that model, in your own words (markdown, not code) explain what the score means
y_pred = model.predict(X_test)
r_squared = r2_score(y_test, y_pred)
print("R² Score:", r_squared)

# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
[coef_temp, coef_kcl] = model.coef_
intercept = model.intercept_
h_x = f"h(x) = {intercept:.5f} + {coef_temp:.5f} T + {coef_kcl:.5f} KCl"
print(h_x)



Predicted Size nm^3 for Temperature = 600 °C & Mols KCL = 500: 6.2664e+05
R² Score: 0.8552472077276095
h(x) = -409391.47958 + 866.14641 T + 1032.69507 KCl


Write the linear equation of a slime: (example equation: $E = mc^2$)

$h(x) = -409391.48 + (866.14641) \cdot T + (1032.69507) \cdot KCl$

Report on score and explain meaning:

The R² score of 0.855 means that 85.5% of the variance in the dependent variable (Size nm^3) can be explained by the independent variables (Temperature °C and Mols KCL). 85.5% indicates a strong performing regression model that accounts for a significant amount of variance in the target variable.

## 4. Use Cross Validation

In [77]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
kf = KFold(n_splits=5, shuffle=True, random_state=42)
cv_scores = cross_val_score(model, x, y, cv=kf)
print("Cross Validation Scores:", cv_scores)
print("Mean CV Score: {:.4f}".format(cv_scores.mean()))
print("Standard Deviation of CV Scores: {:.4f}".format(cv_scores.std()))

# Report on their finding and their significance

Cross Validation Scores: [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
Mean CV Score: 0.8597
Standard Deviation of CV Scores: 0.0184


Write findings here:
The cross validation scores indicate a strong and consistent relationship between the features and labels. The mean is high at 0.8597 and the standard deviation is low at 0.0184. Thus, slime size can confidently be predicted with temperature and mols kcl.

## 5. Using Polynomial Regression

In [87]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)

poly_model = make_pipeline(
    PolynomialFeatures(degree=2), 
    LinearRegression()
)
cv_scores = cross_val_score(poly_model, x, y, cv=kf)
print("Cross Validation Scores:", cv_scores)
print("Mean CV Score: {:.4f}".format(cv_scores.mean()))
print("Standard Deviation of CV Scores: {:.4f}".format(cv_scores.std()))

# Report on the metrics and output the resultant equation as you did in Part 3.

Cross Validation Scores: [1. 1. 1. 1. 1.]
Mean CV Score: 1.0000
Standard Deviation of CV Scores: 0.0000


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

Report on the score and interpret:

The cross validation scores are all 1 with 0 standard deviation. This is unusual and suggests that the model is memorizing the data exactly, there is a data leakage somewhere, or the relationship between the features and label is deterministic. Further contextual analysis of the data is warranted to verify this result.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [108]:
# Load the dataset. Then train and evaluate the classification models.
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier

from sklearn.svm import SVC
df = pd.read_csv('ckd_feature_subset.csv')
random_seed = 42
#print(df.head())

X = df.drop(columns=['Target_ckd'])
y = df['Target_ckd']

models = {
    'Logistic Regression': LogisticRegression(random_state=42),
    'Support Vector Machine': SVC(random_state=42),
    'k-Nearest Neighbors': KNeighborsClassifier(),
    'Neural Network': MLPClassifier(random_state=42, max_iter=2000)
}
results = {}

kf = KFold(n_splits=5, shuffle=True, random_state=42)
for name, model in models.items():
    cv_results = cross_val_score(model, X, y, cv=kf)
    results[name] = {
        'mean accuracy': sum(cv_results) / len(cv_results),
        'std accuracy': (sum((x - (sum(cv_results) / len(cv_results))) ** 2 for x in cv_results) / len(cv_results)) ** 0.5
    }
print(pd.DataFrame(results).T)


                        mean accuracy  std accuracy
Logistic Regression          0.856559      0.066269
Support Vector Machine       0.928172      0.047601
k-Nearest Neighbors          0.927957      0.052440
Neural Network               0.935054      0.040466


Logistic regression:
 - Achieved mean accuracy of 0.856559 and std accuracy of 0.066269. This was the lowest of all the models, but still fairly accurate. Being a linear model, it may not have been able to capture all of the non-linear relationships in the data as effectively as other models.

Support vector machine:
 - Achieved mean accuracy of 0.928172 and std accuracy of 0.047601. This model performed very well, with high accuracy and relatively low variability. The SVM's ability to handle high-dimensional data and find optimal decision boundaries likely contributed to its strong performance.

K-nearest neighbors:
 - Achieved mean accuracy of 0.927957 and std accuracy of 0.052440. The k-NN model also performed well, with accuracy close to that of the SVM model. Its simplicity and effectiveness in capturing local patterns in the data may have contributed to its performance. However, it had slightly higher variability compared to the SVM model.

Neural networks:
 - Achieved mean accuracy of 0.935054 and std accuracy of 0.040466. This was the highest mean accuracy among all the models, indicating that the neural network model performed the best. It also had the lowest standard deviation, suggesting high consistency in its performance. The neural network's ability to capture complex patterns in the data likely contributed to its better performance.

## Results and Conclusion for Classification Experiments

In [107]:
# Experiments with Neural Network.
nn_configs = [
    {'hidden_layer_sizes': (50,), 'max_iter': 2000, 'learning_rate_init': 0.001, 'random_state': 42},
    {'hidden_layer_sizes': (100,), 'max_iter': 2000, 'learning_rate_init': 0.001, 'random_state': 42},
    {'hidden_layer_sizes': (50, 50), 'max_iter': 2000, 'learning_rate_init': 0.001, 'random_state': 42}
]
nn_results = {}

for config in nn_configs:
    model = MLPClassifier(**config)
    cv_results = cross_val_score(model, X, y, cv=kf)
    config_name = f"hidden_layers={config['hidden_layer_sizes']}"
    nn_results[config_name] = {
        'mean accuracy': sum(cv_results) / len(cv_results),
        'std accuracy': (sum((x - (sum(cv_results) / len(cv_results))) ** 2 for x in cv_results) / len(cv_results)) ** 0.5
    }
print(pd.DataFrame(nn_results).T)

                        mean accuracy  std accuracy
hidden_layers=(50,)          0.935054      0.040466
hidden_layers=(100,)         0.935054      0.040466
hidden_layers=(50, 50)       0.954624      0.043696


## Results and Conclusion for Neural Network Experiments

Neural network with hidden_layers=(50,):
 - Achieved mean accuracy of 0.935054 and std accuracy of 0.040466. This configuration performed very well, with high accuracy and low variability. The single hidden layer with 50 neurons was effective in capturing the patterns in the data.

Neural network with hidden_layers=(100,):
 - Achieved mean accuracy of 0.935054 and std accuracy of 0.040466. This configuration also performed very well, with the same accuracy and variability as the previous configuration. Increasing the number of neurons to 100 did not significantly impact the performance.

Neural network with hidden_layers=(50, 50):
 - Achieved mean accuracy of 0.954624 and std accuracy of 0.043696. This configuration had the highest mean accuracy among all the neural network configurations, indicating that the two hidden layers with 50 neurons each were more effective in capturing the complex patterns in the data. However, it had slightly higher variability compared to the single-layer configurations.